ateam_common/src/km_assignment.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright 2024 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <Eigen/Dense>
+#include <limits>
+#include <vector>
+#include <deque>
+#include "ateam_common/km_assignment.hpp"
+
+namespace ateam_common::km_assignment
+{
+const float EPSILON = 1e-10;
+
+const Eigen::MatrixXd make_square_cost_matrix(
+  const Eigen::MatrixXd & matrix)
+{
+  if (matrix.rows() == matrix.cols()) {
+    return matrix;
+  }
+
+  std::size_t new_dim = std::max(matrix.rows(), matrix.cols());
+
+  //  The value of any non-existant cost is 0 (since we are
+  //  trying to find a MAX cost matching)
+  Eigen::MatrixXd new_mat =
+    Eigen::MatrixXd::Constant(new_dim, new_dim, std::numeric_limits<double>::min());
+
+  new_mat.block(0, 0, matrix.rows(), matrix.cols()) = matrix;
+
+  return new_mat;
+}
+
+inline void compute_slack(
+  const int x,
+  std::vector<double> & slack,
+  std::vector<double> & slackx,
+  const Eigen::MatrixXd & cost,
+  const Eigen::VectorXd & lx,
+  const Eigen::VectorXd & ly
+)
+{
+  for (size_t y = 0; y < static_cast<size_t>(cost.cols()); ++y) {
+    if (lx[x] + ly[y] - cost(x, y) < slack[y]) {
+      slack[y] = lx[x] + ly[y] - cost(x, y);
+      slackx[y] = x;
+    }
+  }
+}
+
+std::vector<int> max_cost_assignment(
+  const Eigen::MatrixXd & cost_matrix
+)
+{
+  auto cost = cost_matrix;
+  if (cost_matrix.cols() != cost_matrix.rows()) {
+    cost = make_square_cost_matrix(cost_matrix);
+  }
+  /*
+      The dlib implementation of this algorithm is based on info available below...
+      although the links were broken when I tried them.
+          http://www.math.uwo.ca/~mdawes/courses/344/kuhn-munkres.pdf
+          http://www.topcoder.com/tc?module=Static&d1=tutorials&d2=hungarianAlgorithm
```

Comment:
We can probably omit these since they are indeed long dead links. It's unfortunate because that uwo.ca link seems to be referenced in a lot of places online.